proxy/socks/tcp.go

```diff
@@ -29,38 +26,27 @@ func NewTCPHandler(proxyHost string, proxyPort uint16, fakeDns dns.FakeDns) core
 	return &tcpHandler{
 		proxyHost: proxyHost,
 		proxyPort: proxyPort,
-		conns:     make(map[core.TCPConn]net.Conn, 16),
 		fakeDns:   fakeDns,
 	}
 }
 
-func (h *tcpHandler) fetchInput(conn core.TCPConn, input io.Reader) {
-	// FIXME maybe use a larger buffer?
-	buf := core.NewBytes(core.BufSize) // 2k buf
-
+func (h *tcpHandler) handleInput(conn net.Conn, input io.ReadCloser) {
 	defer func() {
-		h.Close(conn)
-		conn.Close() // also close tun2socks connection here
-		core.FreeBytes(buf)
+		conn.Close()
+		input.Close()
 	}()
-
-	_, err := io.CopyBuffer(conn, input, buf)
-	if err != nil {
-		// log.Printf("fetch input failed: %v", err)
-		return
-	}
+	io.Copy(conn, input)
 }
 
-func (h *tcpHandler) getConn(conn core.TCPConn) (net.Conn, bool) {
-	h.Lock()
-	defer h.Unlock()
-	if c, ok := h.conns[conn]; ok {
-		return c, true
-	}
-	return nil, false
+func (h *tcpHandler) handleOutput(conn net.Conn, output io.WriteCloser) {
```

Comment:
FYI, this prevents TCP half-close behavior and generates errors in every connection.

Once you are done with the write connection, you will close the read connection too. This has two issues:
- An application that half-closes the connection right after it's done writing will not see the full server response
- The other `io.Copy` call will return a "attempt to read from a closed socket" error, which is not good if you are logging errors.

You should instead call `conn.CloseRead()` and `output.CloseWrite()`. `CloseRead` will release the reader resources, while `CloseWrite` will send a FIN to the remote saying you are done writing. Take a look on what I did for outline-ss-server: https://github.com/Jigsaw-Code/outline-ss-server/blob/8972cd64904d445b27bfb7efd083f1d3bc5bbf29/net/net.go#L69

You will need to pass TCPConn objects though, since net.Conn doesn't have the methods we need for proper close behavior.